### PR TITLE
Fixes and enhancements for gizmos

### PR DESCRIPTION
**Pull request by CedricGuillemet**
Many small fixes/enhancements
- https://forum.babylonjs.com/t/boundingboxgizmo-ishovered-always-false/44095
- https://forum.babylonjs.com/t/boundingboxgizmo-scalebox-snapping/44060
- https://forum.babylonjs.com/t/gizmo-position-doesnt-update-when-coordinates-mode-world/44331
- https://forum.babylonjs.com/t/gizmos-misbehave-when-operating-on-child-objects-with-configured-pivot/42313
- https://forum.babylonjs.com/t/unable-to-drag-position-gizmo-after-changing-pivot/44217
- https://forum.babylonjs.com/t/gizmo-location-with-setpivotpoint/41271
- https://forum.babylonjs.com/t/the-gizmo-uses-the-latest-cameras-in-inspector/31221

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14316/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14316/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14316/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14316/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14316/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14316/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14316/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14316/merge#BCU1XR#0

**Review by sebavan — APPROVED** [review]
Approving the code, but hard for me to judge accurately some of the maths :-)

**Comment by CedricGuillemet** [comment]
> Approving the code, but hard for me to judge accurately some of the maths :-)

PR is still a draft. I plan to fix 2 or 3 more bugs :)

**Comment by CedricGuillemet** [comment]
now it's ready for a review @sebavan :)

File: packages/dev/core/src/Gizmos/boundingBoxGizmo.ts [hunk 1]
```diff
@@ -222,6 +229,38 @@ export class BoundingBoxGizmo extends Gizmo implements IBoundingBoxGizmo {
...
+    public get scalingSnapDistance(): number {
+        return this._scalingSnapDistance;
+    }
+
+    /**
+     * Sets rotation snap distance
+     * @param value the rotation snap distance value
+     */
+    public set rotationSnapDistance(value: number) {
+        this._rotationSnapDistance = value;
```
Comment (RaananW) [inline packages/dev/core/src/Gizmos/boundingBoxGizmo.ts:253]: Same :-)

**Review by sebavan — COMMENTED** [review]
I think you also need to remove the _handlePivot function as it is not used anymore and good to merge !!!

File: packages/dev/core/src/Gizmos/boundingBoxGizmo.ts [hunk 2]
```diff
@@ -222,6 +229,38 @@ export class BoundingBoxGizmo extends Gizmo implements IBoundingBoxGizmo {
         return this._scaleDragSpeed;
     }
 
+    /**
+     * Sets scaling snap distance
+     * @param value the scaling snap distance value
+     */
+    public set scalingSnapDistance(value: number) {
+        this._scalingSnapDistance = value;
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Gizmos/boundingBoxGizmo.ts:237]: yes, why not?

File: packages/dev/core/src/Gizmos/boundingBoxGizmo.ts (hunk 1 above)
Comment (CedricGuillemet) [reply packages/dev/core/src/Gizmos/boundingBoxGizmo.ts:253]: fixed as well

File: packages/dev/core/src/Gizmos/gizmo.ts [hunk 3]
```diff
@@ -449,6 +449,28 @@ export class Gizmo implements IGizmo {
...
+                    Matrix.ScalingToRef(transform.scaling.x, transform.scaling.y, transform.scaling.z, scaleMatrix);
+
+                    const rotationMatrix = TmpVectors.Matrix[2];
+                    r.toRotationMatrix(rotationMatrix);
+
+                    const pivotMatrix = transform.getPivotMatrix();
+                    const invPivotMatrix = TmpVectors.Matrix[3];
+                    pivotMatrix.invertToRef(invPivotMatrix);
+
+                    const lm = pivotMatrix.multiply(scaleMatrix).multiply(rotationMatrix).multiply(invPivotMatrix);
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply packages/dev/core/src/Gizmos/gizmo.ts]: fixed

**Review by sebavan — APPROVED** [review]

**Pull request merged**